File: solutions/234.palindrome-linked-list/palindrome-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
const isPalindrome = head => {
  if (!head) return true
  let arr = []

  while (head) {
    arr.push(head.val)
    head = head.next
  }

  return isPali(arr)
}

const isPali = arr => {
  return arr.join('') === arr.reverse().join('')
}

File: solutions/234.palindrome-linked-list/palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        1 2 3 4 5
        1 2 3 4
          | |
        """

        slow = head
        fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        if fast and fast.next is None:
            slow = slow.next

        def reverse(prev, head):
            if not head:
                return prev

            next = head.next
            head.next = prev

            return reverse(head, next)

        ref = reverse(None, slow)

        while ref:
            if ref.val != head.val:
                return False

            ref = ref.next
            head = head.next

        return True
